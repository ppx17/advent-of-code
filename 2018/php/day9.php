<?php
ini_set('memory_limit', '512M');
$data = $data ?? file_get_contents("../input/input-" . basename(__FILE__, '.php') . ".txt");

function rotate(SplDoublyLinkedList $board, $steps) {
    if($steps > 0) {
        for($i=0;$i<$steps;$i++) {
            $board->unshift($board->pop());
        }
    }else{
        for($i=0;$i>$steps;$i--) {
            $board->push($board->shift());
        }
    }
}

function getHighScore(int $numPlayers, int $numMarbles): int
{
    $playerScores = [];
    $board = new SplDoublyLinkedList();
    $board->push(0);
    for ($marble = 1; $marble <= $numMarbles; $marble++) {
        if ($marble % 23 === 0) {
            rotate($board, -7);
            $playerScores[$marble % $numPlayers] += $board->pop() + $marble;
        } else {
            rotate($board, 2);
            $board->push($marble);
        }
    }
    return max($playerScores);
}

preg_match('/(?<players>\d+) players; last marble is worth (?<marbles>\d+) points/', $data, $matches);
echo "Part 1: ".getHighScore($matches['players'], $matches['marbles']).PHP_EOL;
echo "Part 2: ".getHighScore($matches['players'], $matches['marbles'] * 100).PHP_EOL;