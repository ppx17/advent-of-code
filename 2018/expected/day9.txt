Part 1: 390592
Part 2: 3277920293
